src/coreclr/jit/helperexpansion.cpp

```diff
@@ -413,6 +413,388 @@ bool Compiler::fgExpandRuntimeLookupsForCall(BasicBlock** pBlock, Statement* stm
+    //
+    BasicBlock* lengthCheckBb = fgNewBBafter(noSizeCheck ? BBJ_NONE : BBJ_COND, prevBb, true);
+    lengthCheckBb->bbFlags |= BBF_INTERNAL;
+
+    // In 99% cases "this" is expected to be "static readonly UTF8EncodingSealed s_default"
+    // which is a static readonly object that is never null
+    const bool thisIsKnownNonNull =
+        thisArg->gtVNPair.BothEqual() && vnStore->IsKnownNonNull(thisArg->gtVNPair.GetLiberal());
+
+    // Spill all original arguments to locals in the lengthCheckBb to preserve all possible side-effects.
+    thisArg = SpillExpression(this, thisArg, lengthCheckBb, debugInfo);
+    srcPtr  = SpillExpression(this, srcPtr, lengthCheckBb, debugInfo);
+    srcLen  = SpillExpression(this, srcLen, lengthCheckBb, debugInfo);
+    dstPtr  = SpillExpression(this, dstPtr, lengthCheckBb, debugInfo);
+    dstLen  = SpillExpression(this, dstLen, lengthCheckBb, debugInfo);
+
+    if (!noSizeCheck)
+    {
+        // Set bytesWritten -1 by default, if the fast path is not taken we'll return it as the result.
+        GenTree* bytesWrittenDefaultVal = gtNewAssignNode(gtClone(resultLcl), gtNewIconNode(-1));
+        fgInsertStmtAtEnd(lengthCheckBb, fgNewStmtFromTree(bytesWrittenDefaultVal, debugInfo));
+    }
+
+    // We don't need "this" object in the fast path so insert an explicit null check here.
+    // (after we evaluated all arguments)
+    if (!thisIsKnownNonNull)
+    {
+        GenTree* thisNullcheck = gtNewNullCheck(gtClone(thisArg), lengthCheckBb);
+        fgInsertStmtAtEnd(lengthCheckBb, fgNewStmtFromTree(thisNullcheck, debugInfo));
+    }
+
+    if (!noSizeCheck)
+    {
+        GenTree* lengthCheck = gtNewOperNode(GT_LT, TYP_INT, gtClone(dstLen), srcLenCnsNode);
+        lengthCheck->gtFlags |= GTF_RELOP_JMP_USED;
+        Statement* lengthCheckStmt = fgNewStmtFromTree(gtNewOperNode(GT_JTRUE, TYP_VOID, lengthCheck), debugInfo);
+        fgInsertStmtAtEnd(lengthCheckBb, lengthCheckStmt);
+        lengthCheckBb->bbCodeOffs    = block->bbCodeOffsEnd;
+        lengthCheckBb->bbCodeOffsEnd = block->bbCodeOffsEnd;
+    }
+
+    //
+    // Block 2: fastpathBb - unrolled loop that copies the UTF8 const data to the destination
+    //
+    // We're going to emit a series of loads and stores to copy the data.
+    // In theory, we could just emit the const U8 data to the data section and use GT_BLK here
+    // but that would be a bit less efficient since we would have to load the data from memory.
+    //
+    BasicBlock* fastpathBb = fgNewBBafter(BBJ_NONE, lengthCheckBb, true);
+    fastpathBb->bbFlags |= BBF_INTERNAL;
+
+    // The widest type we can use for loads
+    const var_types maxLoadType = roundDownMaxRegSize(srcLenCns);
+
+    // How many iterations we need to copy UTF8 const data to the destination
+    unsigned iterations = srcLenCns / genTypeSize(maxLoadType);
+
+    // Add one more iteration if we have a remainder
+    iterations += srcLenCns % genTypeSize(maxLoadType) == 0 ? 0 : 1;
```

Comment:
```suggestion
    // How many iterations we need to copy UTF8 const data to the destination
    unsigned iterations = srcLenCns / genTypeSize(maxLoadType);

    // Add one more iteration if we have a remainder
    iterations += (srcLenCns % genTypeSize(maxLoadType)) == 0 ? 0 : 1;
```

(or write it in the `(x + (y - 1) ) / y` way, that's usually how we do it in the JIT I think)